src/NuGet.Clients/NuGet.VisualStudio.Common/OutputConsoleLogger.cs

```diff
@@ -104,7 +110,12 @@ public void Log(MessageLevel level, string message, params object[] args)
                     message = string.Format(CultureInfo.CurrentCulture, message, args);
                 }
 
-                NuGetUIThreadHelper.JoinableTaskFactory.Run(() => OutputConsole.WriteLineAsync(message));
+                NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```

Comment:
I'm a bit worried about ordering here. 
This could lead to some messages coming out of order.

Reply:
Just pushed fix for that.

Reply:
The semaphore doesn't guarantee the ordering though does it?

Reply:
Not really, but guarantees no overlaps. However, we may go one further, and institute a queue (we have one in deployment service), so that logging is fully async, non-blocking, and in guaranteed order. I will send update a bit later.

Reply:
Assuming two logging operations are coming from same context (same background thread), acquisition of semaphore is happening on exactly same context (in same thread), before attempt to switch to foreground thread to output messages.

So, if you have two successive logging operations, first is guaranteed to acquire semaphore before second one. So that's solved as-is.

However, that also does not provide full potential performance benefits, as semaphore is held until logging operation is completed, so subsequent logging operations are still serialized (non-blockingly, though).

Reply:
> The semaphore doesn't guarantee the ordering though does it?

Yes, `ReentrantSemaphore` effectively has a FIFO queue of folks waiting for the semaphore.